assets/js/sections/contesting.js

```diff
@@ -208,26 +213,43 @@ $('#start_date').change(function () {
 
 // On Key up check and suggest callsigns
 $("#callsign").keyup(function () {
-	var call = $(this).val();
+	let call = $(this).val().toUpperCase();
 	if (call.length >= 3) {
 
-		$.ajax({
-			url: 'lookup/scp',
-			method: 'POST',
-			data: {
-				callsign: $(this).val().toUpperCase()
-			},
-			success: function (result) {
-				$('.callsign-suggestions').text(result);
-				highlight(call.toUpperCase());
-			}
-		});
+        if ( scp_data.request == "" || ! call.startsWith(scp_data.request) ) {
+          scp_data.request = call;
+          scp_data.data = [];
+          $.ajax({
+              url: 'lookup/scp',
+              method: 'POST',
+              data: {
+                  callsign: call
+              },
+              success: function (result) {
+                  var call_now = $("#callsign").val().toUpperCase(); // Might have changed while running the query...
```

Comment:
1) I still dont see the benefit of updating this since the returned data in `result` will potentially have nothing to do with the updated call sign.  I think a more nuanced approach might be to cancel the existing ajax request if the user changes the input before the request has completed.

2)  I still don't understand why we have `scp_data.request == ""` check? It only return `true` the first request, i dont see the benefit of it ever being `false`?

3) should we go ahead and update the `startsWith` to `includes`?
 

I havent tested it, but i think something like this should work:

```javascript
$("#callsign").keyup(function () {
    let call = $(this).val().toUpperCase();
    if (call.length >= 3) {
        // Check if a previous request exists and abort it
        if(scp_data.previousRequest) {
            scp_data.previousRequest.abort();
        }

        if (scp_data.request == "" || !call.startsWith(scp_data.request)) {
            scp_data.request = call;
            scp_data.data = [];
            // Store the current request in scp_data
            scp_data.previousRequest = $.ajax({
                url: 'lookup/scp',
                method: 'POST',
                data: {
                    callsign: call
                },
                success: function (result) {
                    // When the request is complete, nullify the previousRequest to clean up
                    scp_data.previousRequest = null;
                }
            });
        }
    }
});
```

Reply:
1.) You type "2M0" and the request is fired. While the request takes place, you enter "S", and as "2M0S" starts with "2M0" no new request is fired. The time the success-callback is executed `call_now="2M0S"` and the correct subset of the delivered data is shown.

2.) Every string starts with "" (try it out ;) ), so we need a starting point. An alternative would be to initialize `scp_data.request` with some magic value...

3.) Why?

4.) `scp` expects '0' in the query, but returns 'Ø' in the result (have a look in the code). Therefore this should be right - maybe @magicbug can comment on this.

Code example: Sure, there are many ways to go.

Reply:
1) what if i quickly press ctrl + a to select everything and start typing `WA6`?
2) I dont understand, a string cant start with an empty string.   `scp_data.request == ""` is specifically checking if that string is empty, not if its undefined, not if its false, only empty string.
3) I played with it, you're right..dont need it.
4) If that's the case, why do we only have the logic on QSO.js and not contest.js?

Reply:
1.) I'd suggest to try it out or to think about what-happens-when in the code, like I did in the previous example. Let me know if I can assist you, but maybe a call would be better then :).
2.) Did you try `"anything".startsWith("")` in a JS console?
4.) What differences do you mean? In dev-Branch?

Reply:
1) Ok, so if a user changes their input completely before the results have returned it fires off a second request, leaving 2 requests open to the server at a single time.

See here:
![image showing stale requests being made unnecessarily](https://github.com/magicbug/Cloudlog/assets/6586559/fc00c78e-3673-4530-87f0-0c2f0728b1e2)

It can be improved so we cancel the previous request instead of letting it finish:

![image showing us abort from a stale request](https://github.com/magicbug/Cloudlog/assets/6586559/bcfd7a16-e880-4f94-a77f-011f8e60079b)


2) Played with and It seems to only handle the first time the event is fired, which is fine...
4) Ahh i see it was added in both spots now

I went ahead and fired up the branch and tested the code below, i think the following implementation is easier to reason about, adds some performance upgrades, and updates it to modern javascript patterns.

- We really dont need to redeclare `call` as `call_now`  in the success call back, below is a cleaned up version showing you dont need it
- use `===` in the comparison
- use const and keep variables immutable
- adds abort to the request if the user changes the input based on the start of the string only

```javascript
// On Key up check and suggest callsigns
$("#callsign").keyup(function () {
	const call = $(this).val().toUpperCase();
	const cleanedCall = call.replace('0','Ø');

	// Exit early if the users starts typing a completely new call
    if(scp_data.previousRequest && !call.startsWith(scp_data.request)) {
        scp_data.previousRequest.abort();
    }

	
	if (call.length >= 3) {
        if ( scp_data.request === "" || !call.startsWith(scp_data.request) ) {
          scp_data.request = call;
          scp_data.data = [];
          scp_data.previousRequest = $.ajax({
              url: 'lookup/scp',
              method: 'POST',
              data: {
                  callsign: call
              },
              success: function (result) {
                if (call.startsWith(call)) {
                	scp_data.data = result.split(" ");
                    $('.callsign-suggestions').text(filterCallsignList(cleanedCall, scp_data.data));
                    highlight(cleanedCall);
                  }
              },
			  complete: function () {
				scp_data.previousRequest = null;
			  }
          });

        } else {
          // Filter the already obtained list:
          $('.callsign-suggestions').text(filterCallsignList(cleanedCall, scp_data.data));
          highlight(cleanedCall);

        }
		// moved to blur
		// checkIfWorkedBefore();
		var qTable = $('.qsotable').DataTable();
		qTable.search(call).draw();
	}
	else {
		$('.callsign-suggestions').text("");
	}
});
```

Reply:
You are again missing out that by the time, the success-callback is executed, the user might have typed additional letters.

I see no reason for complicating the code by introducing more global-scope items.

Reply:
ope, i see now about `call` `call_now`, updating code above in a sec

Reply:
oh no, this isnt an issue with that at all...it seems like the caching isnt working like it was thought to?


When i did a search for "WA6BND" it grabbed "WA6" whose results did not include `WA6BND`, heres the list it did return:

```
HP3/WA6YJW VP2V/WA6O WA6AAC WA6AEE WA6ALA WA6AM WA6AMK WA6APQ WA6AQL WA6ARA WA6AUP WA6AZP WA6BEF WA6BFG WA6BGS WA6BJH WA6BJM WA6BKR WA6BXI WA6BXV WA6CAL WA6CDO WA6CGR WA6CQE WA6D WA6DBC WA6DH WA6DIL WA6DIR WA6DKD WA6DKN WA6DNR WA6DON WA6DPP WA6EBO WA6ECH WA6ED WA6EEM WA6EGG WA6EJO WA6ERA WA6EZV WA6F WA6FFJ WA6FGV WA6FIZ WA6FV WA6FWN WA6FXT WA6FXY WA6FYD WA6FYV WA6FZV WA6G WA6GBW WA6GD WA6GDU WA6GFE WA6GFR WA6GJW WA6GUN WA6GVI WA6GXQ WA6HNA WA6HQQ WA6HTP WA6HXD WA6HXM WA6HXW WA6HZT WA6HZY WA6I WA6IBU WA6IKE WA6IOQ WA6IPD WA6IQO WA6J WA6JA WA6JBZ WA6JGM WA6JJA WA6JJM WA6JKN WA6JRP WA6JRZ WA6KFI WA6KHK WA6KJN WA6KJV WA6KLK WA6KSF WA6KTK WA6KYR WA6LAU WA6LBU WA6LDI WA6LGS WA6LIE WA6LJS WA6LOC WA6LOS/DU3 WA6MBL WA6MCL WA6MDI WA6MEM WA6MHZ WA6MIZ WA6MM WA6MNV WA6MOW WA6MP WA6MPN WA6MQF WA6MRK WA6MTN WA6MTZ WA6MVN WA6N WA6NEA WA6NEQ WA6NFA WA6NFJ WA6NHC WA6NKM WA6NPA WA6NPC WA6NUO WA6NVL WA6O WA6OC WA6OCV WA6OCV/M WA6OEF WA6OEM WA6OIB WA6OJY WA6OSX WA6OT WA6OUD WA6OVP WA6OWM WA6OWM/KH6 WA6OWM/QRP WA6OXK WA6P WA6PEL WA6PHR WA6PIC WA6PKB WA6POZ WA6PPG WA6PPX WA6PSI WA6PWP WA6PXX WA6PY WA6Q WA6QBU WA6QGH WA6QHD WA6QMQ WA6QYS WA6RE WA6RKK WA6RLR WA6RSV WA6RWI WA6RYA WA6RZI WA6RZR WA6SCW WA6SDB WA6ST WA6STV WA6SWM WA6SXU WA6SZE WA6T WA6TE WA6TFZ WA6TGT WA6TLA WA6TLL WA6TMJ WA6TOW WA6TQJ WA6TQT WA6TVO WA6TWF WA6TWJ WA6TXU WA6TYQ WA6UIM WA6UMA WA6URY WA6UVW WA6UXA WA6UYO WA6VAB WA6VHU WA6VPJ WA6VWW WA6WA WA6WGS WA6WJN WA6WPG WA6WXM WA6YAW WA6YBR WA6YDQ WA6YEE WA6YOU WA6YPE WA6YSM WA6ZKM WA6ZTY WA6Ã˜87SWL YU/WA6MP
```


So, this is a huuuuge reason to implement debounce if you want caching :)

Reply:
Ok, so they're not global scoped, they're scoped to the anonymous function that is passed into the event.
I've gone ahead and updated it to catch the users input if its changed after the results are returned.
This is a much cleaner implementation, please use it:

```javascript
// On Key up check and suggest callsigns
$("#callsign").keyup(function () {
	const call = $(this).val().toUpperCase();
	const cleanedCall = call.replace('0', 'Ø');

	// Exit early if the users starts typing a completely new call
	if (scp_data.previousRequest && !call.startsWith(scp_data.request)) {
		scp_data.previousRequest.abort();
	}

	if (call.length >= 3) {
		if (scp_data.request === "" || !call.startsWith(scp_data.request)) {
			scp_data.request = call;
			scp_data.data = [];
			scp_data.previousRequest = $.ajax({
				url: 'lookup/scp',
				method: 'POST',
				data: {
					callsign: call
				},
				success: function (result) {
					const currentCall = $("#callsign").val().replace('0', 'Ø').toUpperCase();
					if (currentCall.startsWith(call)) {
						scp_data.data = result.split(" ");
						$('.callsign-suggestions').text(filterCallsignList(currentCall, scp_data.data));
						highlight(currentCall);
					}
				},
				complete: function () {
					scp_data.previousRequest = null;
				}
			});

		} else {
			// Filter the already obtained list:
			$('.callsign-suggestions').text(filterCallsignList(cleanedCall, scp_data.data));
			highlight(cleanedCall);
		}
		var qTable = $('.qsotable').DataTable();
		qTable.search(call).draw();
	}
	else {
		$('.callsign-suggestions').text("");
	}
});
```

Reply:
I don't see an advantage but increased complexity, sorry.

Scoping: I guess we are not talking about the same thing.

Reply:
Did you try the code?  Which part is too complex?  

Also, You clearly don't understand scoping.

Reply:
Ok, the whole pr increases complexity, but 3 lines that suggested is  too complex?  

Bro, I've been trying to be nice this whole pr.  Your pr was messy, used old and outdated JavaScript patterns.   I'm not saying this to be mean, but to keep the code quality high for a project I love.